From dc6c992604fb03163d8957e288cf5fd32c868fd6 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:04:19 -0400
Subject: [powerpc] perf: increase the perf HW events to 6

Message-id: <1410545655-205645-31-git-send-email-dzickus@redhat.com>
Patchwork-id: 94533
O-Subject: [RHEL7 PATCH 030/626] powerpc/perf: increase the perf HW events to 6
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 96c3c9e78f7fe45303985cce13ce26e35afd01ba
Author: Catalin Udma <catalin.udma@freescale.com>
Date:   Wed Jun 5 15:22:08 2013 -0500

    powerpc/perf: increase the perf HW events to 6

    This change is required after the e6500 perf support has been added.
    There are 6 counters in e6500 core instead of 4 in e500 core and
    the MAX_HWEVENTS counter should be changed accordingly from 4 to 6.
    Added also runtime check for counters overflow.

    Signed-off-by: Catalin Udma <catalin.udma@freescale.com>
    Signed-off-by: Lijun Pan <Lijun.Pan@freescale.com>
    Signed-off-by: Scott Wood <scottwood@freescale.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/perf_event_fsl_emb.h b/arch/powerpc/include/asm/perf_event_fsl_emb.h
index 718a9fa..a581654 100644
--- a/arch/powerpc/include/asm/perf_event_fsl_emb.h
+++ b/arch/powerpc/include/asm/perf_event_fsl_emb.h
@@ -13,7 +13,7 @@
 #include <linux/types.h>
 #include <asm/hw_irq.h>
 
-#define MAX_HWEVENTS 4
+#define MAX_HWEVENTS 6
 
 /* event flags */
 #define FSL_EMB_EVENT_VALID      1
diff --git a/arch/powerpc/perf/core-fsl-emb.c b/arch/powerpc/perf/core-fsl-emb.c
index 106c533..0b13f74 100644
--- a/arch/powerpc/perf/core-fsl-emb.c
+++ b/arch/powerpc/perf/core-fsl-emb.c
@@ -462,6 +462,12 @@ static int fsl_emb_pmu_event_init(struct perf_event *event)
  int num_restricted;
  int i;
 
+ if (ppmu->n_counter > MAX_HWEVENTS) {
+  WARN(1, "No. of perf counters (%d) is higher than max array size(%d)\n",
+   ppmu->n_counter, MAX_HWEVENTS);
+  ppmu->n_counter = MAX_HWEVENTS;
+ }
+
  switch (event->attr.type) {
  case PERF_TYPE_HARDWARE:
   ev = event->attr.config;
-- 
1.7.1